quinn-proto/src/connection/mod.rs

```diff
@@ -2362,7 +2386,7 @@ impl Connection {
             Header::Retry {
                 src_cid: rem_cid, ..
             } => {
-                if self.side.is_server() {
+                if self.side_state.side().is_server() {
```

Comment:
Unfortunately that causes borrowing errors, because we need a mutable reference to the token later, but in the mean time we call methods on `self`.

<details><summary>Attempted diff</summary>

```diff
diff --git a/quinn-proto/src/connection/mod.rs b/quinn-proto/src/connection/mod.rs
index fcc0d4a5..a3b631e4 100644
--- a/quinn-proto/src/connection/mod.rs
+++ b/quinn-proto/src/connection/mod.rs
@@ -2442,9 +2442,9 @@ impl Connection {
             Header::Retry {
                 src_cid: rem_cid, ..
             } => {
-                if self.side_state.is_server() {
+                let SideState::Client { token, .. } = &mut self.side_state else {
                     return Err(TransportError::PROTOCOL_VIOLATION("client sent Retry").into());
-                }
+                };
 
                 if self.total_authed_packets > 1
                             || packet.payload.len() <= 16 // token + 16 byte tag
@@ -2500,9 +2500,6 @@ impl Connection {
                 self.streams.retransmit_all_for_0rtt();
 
                 let token_len = packet.payload.len() - 16;
-                let SideState::Client { ref mut token, .. } = self.side_state else {
-                    unreachable!("we already short-circuited if we're server");
-                };
                 *token = packet.payload.freeze().split_to(token_len);
                 self.state = State::Handshake(state::Handshake {
                     expected_token: Bytes::new(),

```

</details>

<details><summary>Resultant errors</summary>

```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
    --> quinn-proto/src/connection/mod.rs:2476:21
     |
2445 |                 let SideState::Client { token, .. } = &mut self.side_state else {
     |                                                       -------------------- first mutable borrow occurs here
...
2476 |                     self.on_packet_acked(now, 0, info);
     |                     ^^^^ second mutable borrow occurs here
...
2503 |                 *token = packet.payload.freeze().split_to(token_len);
     |                 ------ first borrow later used here

error[E0499]: cannot borrow `*self` as mutable more than once at a time
    --> quinn-proto/src/connection/mod.rs:2479:17
     |
2445 |                 let SideState::Client { token, .. } = &mut self.side_state else {
     |                                                       -------------------- first mutable borrow occurs here
...
2479 |                 self.discard_space(now, SpaceId::Initial); // Make sure we clean up after any retransmitted Initials
     |                 ^^^^ second mutable borrow occurs here
...
2503 |                 *token = packet.payload.freeze().split_to(token_len);
     |                 ------ first borrow later used here

error[E0502]: cannot borrow `self.side_state` as immutable because it is also borrowed as mutable
    --> quinn-proto/src/connection/mod.rs:2481:69
     |
2445 |                 let SideState::Client { token, .. } = &mut self.side_state else {
     |                                                       -------------------- mutable borrow occurs here
...
2481 |                     crypto: Some(self.crypto.initial_keys(&rem_cid, self.side_state.side())),
     |                                                                     ^^^^^^^^^^^^^^^ immutable borrow occurs here
...
2503 |                 *token = packet.payload.freeze().split_to(token_len);
     |                 ------ mutable borrow later used here

error[E0499]: cannot borrow `*self` as mutable more than once at a time
    --> quinn-proto/src/connection/mod.rs:2497:21
     |
2445 |                 let SideState::Client { token, .. } = &mut self.side_state else {
     |                                                       -------------------- first mutable borrow occurs here
...
2497 |                     self.remove_in_flight(pn, &info);
     |                     ^^^^ second mutable borrow occurs here
...
2503 |                 *token = packet.payload.freeze().split_to(token_len);
     |                 ------ first borrow later used here
```

</details>